Allow apt mirrors

diffway opened the pull request.

diffway committed:
allow use of local mirrors

diffway committed:
Merge branch 'progmaticltd:main' into allow-apt-mirrors

diffway committed:
Merge branch 'progmaticltd:main' into allow-apt-mirrors

progmaticltd commented:
Again, thanks for this MR, that could be useful to many users. 

Would you mind extending this idea a little, please ?

- If we want to let the user replace `deb.debian.org` by another mirror, then we need to organise this differently, IMHO. For instance, by creating a `mirror` variable, with the default being `deb.debian.org`, and then, allow the users to override them.
- Second, as this is a very sensitive value, then I would like to check that the value is valid before deploying the sources.list file, for instance by downloading first the release file (e.g. `https://debian.mirror.serversaustralia.com.au/debian/dists/bookworm/Release`)
- Last but not least, documentation will need to mention this at least.

Again, thanks for your idea, I never thought about this.

diffway committed:
allow choose apt mirrors

diffway commented:
1. Added `mirror` variable to 'system settings' of each `config/defaults/version-*.yml` file, added discriptive comment.

2. My skills at `ansible` are not very strong, so maybe checking code could be done more efficiently, but it passes:
```sh
ROLE=bootstrap ansible-playbook -e version=small check.yml
```
```
...
TASK [bootstrap : Initialise and update the packages repository] ***********************************
included: ~/Projects/homebox-all/homebox/roles/bootstrap/tasks/check/apt.yml for homebox

TASK [bootstrap : Check apt mirror is valid for us] ************************************************
ok: [homebox] => (item=bookworm)
ok: [homebox] => (item=bookworm-updates)
ok: [homebox] => (item=bookworm-backports)

TASK [bootstrap : Check if we can run apt update without error] ************************************
...
```

3. Added to section [System settings](https://homebox.readthedocs.io/en/main/30-define-your-config/#create-the-blank-configuration)  which is IMO self descriptive. Also I'am not sure how to rebuild `readthedocs`.

progmaticltd commented:
Thank you!
Sorry for the late answer, I will check this asap.

progmaticltd commented on roles/bootstrap/tasks/install/apt.yml:
I was suggesting a pre-test before running the playbook, but checking in Ansible is good as well, goo idea.

I usually tend to restrict the usage of `shell` when there is no corresponding module to do the same with a native module.

Here a version using the ansible uri module:

```yml
- name: Check apt mirror is valid for us
  register: mirror
  ansible.builtin.uri:
    url: http://{{ system.apt.mirror }}/debian/dists/{{ distribution_release }}/Release
    return_content: true
  loop:
    - "{{ system.release }}"
    - "{{ system.release }}-updates"
    - "{{ system.release }}-backports"
  loop_control:
    loop_var: distribution_release
```

You can forget about the content checking using regex, as the content will be thoroughly checked by `apt update` in the following tasks.

The purpose of the test is more to be sure the mirror is reachable from the server.

progmaticltd commented on roles/bootstrap/vars/main.yml:
```diff
@@ -6,3 +6,4 @@ whitelisted_hosts:
   - deb.debian.org
```
We then need to remove this line.

diffway committed:
allow choose apt mirrors, use native ansible to check mirror availability.

diffway commented:
I beleve I've implemented everything as discussed.

arodier commented:
Hello, diffway, sorry for the late answer.
Thank you very much, I am testing it right now, and I will come back to you.
The only issue I found so far, is the fact that `https` does not work with all the mirrors.

arodier commented:
Actually, don't do anything, I can automatically detect if the mirror is only compatible with http and write the template accordingly.

diffway commented:
@arodier could you kindly make a decision about those conflicts?
I honestly doesn't understand what I'm supposed to do:

the file `docs/30-define-your-config.md` is gone in `progmaticltd:dev`
the file `roles/bootstrap/vars/main.yml` looks fine to me.

arodier commented:
Sorry, I neglected this issue, I will come back to it in a few days, during holidays.

diffway commented:
Closing this as stale.
